### DISCUSSION TITLE
Unexplainable pitch measurement behaviour

### DISCUSSION THREAD

**Discussion by Devilly** [pybricks/support#1194] [Q&A] [answered]
I've got a piece of code that does nothing more than continuously reading the pitch value of a Technic Hub.

```python
from pybricks.hubs import TechnicHub
from pybricks.pupdevices import Motor, UltrasonicSensor
from pybricks.parameters import Button, Color, Direction, Port, Side, Stop
from pybricks.geometry import Axis
from pybricks.robotics import DriveBase
from pybricks.tools import wait, StopWatch

from usys import stdin, stdout

hub = TechnicHub(Axis.X, Axis.Z)
hub.light.on(Color.GREEN)

motor_left = Motor(Port.A, Direction.COUNTERCLOCKWISE)
motor_right = Motor(Port.B)

motors = [
    motor_right,
    motor_left
]

while(True):
    wait(500)

    tilt = hub.imu.tilt()[0]

    print(str(tilt))

    # for motor in motors:
    #     motor.run(1000)
```
An example output of this when putting the hub against a static object is this:

`-47
-47
-47
-47
-47
-47
-47
-47
-47
-47
-47
-47
-47
-47`

This is exactly what I would expect. A constant value. When I then enable the motor lines of code and run again:

`-47
-30
-39
-39
-44
-45
-45
-31
-50
-52
-46
-46
-39`

The values go all over the place. The motors have nothing attached to them so the hub stays in the exact same position as before.

Any idea what's going on or what I can test more to find out the cause?

**Comment by laurensvalk** [comment]
Could this be vibrations? What do you get if you hold the motor in your hands, not directly against the hub?

If so, the we can probably make it less sensitive to this by incorporating the gyro information. We currently only use the accelerometer for tilt.

**Reply by Devilly** [reply]
@laurensvalk I did two more tests, both in hand and the first without and the second with the motors enabled:

`-18
-19
-19
-19
-19
-19
-19
-19
-19
-19
-19
-19
-20
-19
-19
-19
-19`

and

`-20
-12
-10
-6
-14
-10
-24
-13
-22
-30
-1
-17
-29
-26`

Would it help if I test the same with a Robot Inventor brick?

**Reply by laurensvalk** [reply]
I would be curious to see the results with the hub sitting steadily on the table while you hold the motor in your hand (with the cable plugged in but otherwise loose) so it can't make the hub vibrate.

Just to assess if the nature is mechanical or electrical.

**Reply by Devilly** [reply, marked as answer]
Here you are:

`-27
-27
-27
-27
-27
-27
-27
-27
-27
-26`

Quite steady it is!

I find it fascinating that those small vibrations cause that large of changes in the angle measurements. I was trying to use the pitch for making a balancing bot with PID algorithm but it didn't really seem to behave as I wanted. This may very well be a big influencer on the malbehaviour.

**Reply by Devilly** [reply]
@laurensvalk you've marked above as the answer but it feels to me the behavior is a bug. The angle of the hub factually does not change (this much) based on the vibrations. Can this be fixed?

I'm also curious whether the original Powered Up implementation has the same behavior. May test that, as I can't really imagine it being so.

**Reply by laurensvalk** [reply]
The question was about inexplicable behavior, which now appears explained. Feel free to open a dedicated issue for using the gyro to improve the pitch value :)

**Reply by Devilly** [reply]
Done: https://github.com/pybricks/support/issues/1198
